### ISSUE TITLE
NativeCommandExitException should display "negative" exit codes in hex, not decimal

### ISSUE THREAD

**Issue by sba923** [PowerShell/PowerShell#19886]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

When a native command exits with a "negative" (high-order bit set) exit code, PowerShell 7.4.0-preview.4 displays something like:

`NativeCommandExitException: Program "winget.exe" ended with non-zero exit code: -1978335212.`

This negative value cannot (really) be made sense of unless you convert it to hex (`0x8a150014` in that case).

### Expected behavior

```console
NativeCommandExitException: Program "winget.exe" ended with non-zero exit code: 0x8a150014.
```

### Actual behavior

```console
NativeCommandExitException: Program "winget.exe" ended with non-zero exit code: -1978335212.
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.4.0-preview.4
PSEdition                      Core
GitCommitId                    7.4.0-preview.4
OS                             Microsoft Windows 10.0.19045
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by 237dmitry** [comment]
> non-zero exit code: -1978335212

This is the strange `$LASTEXITCODE` (on linux):
```powershell
PS > bash -c "exit -1978335212"
PS > $LASTEXITCODE             
20
```

**Comment by sba923** [comment]
> > non-zero exit code: -1978335212
> 
> This is the strange `$LASTEXITCODE` (on linux):
> 
> ```powershell
> PS > bash -c "exit -1978335212"
> PS > $LASTEXITCODE             
> 20
> ```

My guess is that on Linux the `$LASTEXITCODE` is the command's exit code `& 0x000000FF`

**Comment by rhubarb-geek-nz** [comment]
https://learn.microsoft.com/en-us/dotnet/api/system.management.automation.nativecommandexitexception.exitcode?view=powershellsdk-7.3.0#system-management-automation-nativecommandexitexception-exitcode

The ExitCode in the exception is an int, hence 32 bit signed integer.

**Comment by 237dmitry** [comment]
> The ExitCode in the exception is an int, hence 32 bit signed integer.

Yes. This is integer but not correct value:

```powershell
PS > bash -c "exit 100000"                                              
PS > ($lec = $LASTEXITCODE).GetType()                                   

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     True     Int32                                    System.ValueType

PS > $lec                                                               
160
```

**Comment by sba923** [comment]
> https://learn.microsoft.com/en-us/dotnet/api/system.management.automation.nativecommandexitexception.exitcode?view=powershellsdk-7.3.0#system-management-automation-nativecommandexitexception-exitcode
> 
> The ExitCode in the exception is an int, hence 32 bit signed integer.
> 
> Also there is no rule that says an exit code should be interpreted as an HRESULT ( which is also a signed 32 bit integer )

There is also no rule that says that messages towards users should make their lives more difficult 😜

**Comment by rhubarb-geek-nz** [comment]
> > The ExitCode in the exception is an int, hence 32 bit signed integer.
> 
> Yes. This is integer but not correct value:
> 
> ```powershell
> PS > bash -c "exit 100000"                                              
> PS > ($lec = $LASTEXITCODE).GetType()                                   
> 
> IsPublic IsSerial Name                                     BaseType
> -------- -------- ----                                     --------
> True     True     Int32                                    System.ValueType
> 
> PS > $lec                                                               
> 160
> ```

Certainly, POSIX has its own encoding for exit codes and typically only supports 8 bits from a clean exit of a process

**Comment by sba923** [comment]
> > The ExitCode in the exception is an int, hence 32 bit signed integer.
> 
> Yes. This is integer but not correct value:
> 
> ```powershell
> PS > bash -c "exit 100000"                                              
> PS > ($lec = $LASTEXITCODE).GetType()                                   
> 
> IsPublic IsSerial Name                                     BaseType
> -------- -------- ----                                     --------
> True     True     Int32                                    System.ValueType
> 
> PS > $lec                                                               
> 160
> ```

again, the value is AND-ed with `0xFF`:

```powershell
pwsh> 100000 -band 255
160
```

To be reported in another issue.

**Comment by rhubarb-geek-nz** [comment]
> pwsh> 100000 -band 255
> 160
> ```
> 
> To be reported in another issue.

That would be correct, 100000 is 0x186A0, and 0xA0 is 160. POSIX only passes the lower 8 bits of the exit value when a process exits cleanly.

**Comment by 237dmitry** [comment]
> again, the value is AND-ed with 0xFF:

Then why (bash):
```
$ bash -c "exit 1000000"
$ echo -e $?
64
```

**Comment by sba923** [comment]
> > pwsh> 100000 -band 255
> > 160
> > ```
> > 
> > To be reported in another issue.
> > ```
> 
> That would be correct, 100000 is 0x186A0, and 0xA0 is 160. POSIX only passes the lower 8 bits of the exit value when a process exits cleanly.

so there's no PowerShell issue there.

**Comment by sba923** [comment]
Back to my issue / change suggestion?

**Comment by sba923** [comment]
> > again, the value is AND-ed with 0xFF:
> 
> Then why (bash):
> 
> ```
> $ bash -c "exit 1000000"
> $ echo -e $?
> 64
> ```

because `1000000` is `0x000F4240` and `0x40` is `64`

**Comment by sba923** [comment]
> > Back to my issue / change suggestion?
> 
> The message is completely correct, the value is a signed integer, and there is no rule that says it should be interpreted as an HRESULT

I never suggested to interpret is an `HRESULT`.

I just said that understanding the value (looking it up on the 'net, in .h files, whatever) is _way_ easier with the hex representation than with the huge negative decimal number.

**Comment by rhubarb-geek-nz** [comment]
> I just said that understanding the value (looking it up on the 'net, in .h files, whatever) is _way_ easier with the hex representation than with the huge negative decimal number.

That may be so but you are making assumptions about the value of ExitCode that are not always true. The value of the exit code is entirely at the whim of the process that has exited, whether it was a value from GetLastError, or an HRESULT or a custom value specific to the application. The only convention that exists is that if the value is zero then it is reporting no error.

**Comment by rhubarb-geek-nz** [comment]
Also, if you take a negative 32 bit value and convert it to hex you have effectively now made it a positive number which now no longer matches the original exit code. 

0x8a150014 is not the same as -1978335212, it is actually 2316632084.

**Comment by sba923** [comment]
> Also, if you take a negative 32 bit value and convert it to hex you have effectively now made it a positive number which now no longer matches the original exit code.
> 
> 0x8a150014 is not the same as -1978335212, it is actually 2316632084.

![image](https://github.com/PowerShell/PowerShell/assets/12860484/8ca8877f-9528-441a-a212-08751f80c796)

**Comment by rhubarb-geek-nz** [comment]
> 0x8a150014

```
PS > [Int64]::Parse('8a150014','HexNumber')
2316632084
```

**Comment by 237dmitry** [comment]
> because 1000000 is 0x000F4240 and 0x40 is 64

Okay, everything topsy-turvy. Like `UnixStat`.

**Comment by sba923** [comment]
> PS > [Int64]::Parse('8a150014','HexNumber')
> 2316632084
> 

Above, you wrote:

> https://learn.microsoft.com/en-us/dotnet/api/system.management.automation.nativecommandexitexception.exitcode?view=powershellsdk-7.3.0#system-management-automation-nativecommandexitexception-exitcode
> 
> The ExitCode in the exception is an int, hence 32 bit signed integer.

So, please be consistent with yourself 😜

**Comment by rhubarb-geek-nz** [comment]
Yes, but if you printed it in the message that context is lost. If I had an app that exited with -1, -2, -3 to indicate status, your proposal would then change them to 0xFFFFFFFF, 0xFFFFFFFE, 0xFFFFFFFD even though I had no intention of looking them up in header files.

**Comment by kasini3000** [comment]
```
#rocky linux 9.2
PS /root> bash -c "exit -2"  
PS /root> $LASTEXITCODE    
254
PS /root> bash -c "exit -1"
PS /root> $LASTEXITCODE    
255
```

Powershell is commendable for allowing us to use exit -1, and the [int] range is greater than [0--255], which is sufficient.

**Comment by kasini3000** [comment]
```
#rocky linux 9.2
PS /root> bash -c "exit -2"  
PS /root> $LASTEXITCODE    
254
PS /root> bash -c "exit -1"
PS /root> $LASTEXITCODE    
255
```

Powershell is commendable for allowing us to use exit -1, and the [int] range is greater than bash [0--255], which is sufficient.

**Comment by sba923** [comment]
> Yes, but if you printed it in the message that context is lost. If I had an app that exited with -1, -2, -3 to indicate status, your proposal would then change them to 0xFFFFFFFF, 0xFFFFFFFE, 0xFFFFFFFD even though I had no intention of looking them up in header files. Might I suggest that somebody with sufficient knowledge to know where to find error codes in header files might also be able to write a PowerShell script to deconstruct HRESULT components from an exit code that was printed in decimal.

My 35 years of experience have seen, *statistically*, tons of `0x8xxxxxxx` or `0xCxxxxxxx` "exit codes" (often, unhandled exceptions, last week's example was a `0xC0000374` i.e. `STATUS_HEAP_CORRUPTION`), and very few, if any, codes in the range -1....-100 or even -1...-1000

**Comment by kilasuit** [comment]
Does this apply to other tools or is this specific to winget?
If other tools can we get some more examples where that tool is doing this too?

Optionally, there is the [Resolve-HResult function as part of PSCX ](https://github.com/Pscx/Pscx/blob/698efdf0ba9cb29b326eb93e4a25ac841cc302dd/Src/Pscx/Modules/Utility/Pscx.Utility.psm1#L682C1-L705C2) that allows you to resolve the output that winget has provided to a HResult

I think this should ideally reported to winget to see if they'd be willing to change their output so that it was a HResult by default.

Hope that helps at least for now.

Will however pass this onto the Engine WG to see if this is something that is sensible to look at doing ourselves or not - though I'd think it perhaps isn't

**Comment by sba923** [comment]
Even though `winget` does document those [error codes](https://github.com/microsoft/winget-cli/blob/master/doc/windows/package-manager/winget/returnCodes.md)--and they're not `HRESULT`s and I see no reason to change that--, any Windows executable that would ungracefully exit from an unhandled exception (such as the tool I was troubleshooting last week, that died with `0xC0000374` i.e. `STATUS_HEAP_CORRUPTION`) will cause the message to include a large negative number that the user *will* have to convert to hex.

**Comment by Luiz-Monad** [comment]
There's no true correct answer for this situation, really. 

So I propose: 

```NativeCommandExitException: Program "winget.exe" ended with non-zero exit code: -1978335212 (0x8a150014).```

Why not both !
![image](https://github.com/PowerShell/PowerShell/assets/4574729/7be450e4-2c2c-4ca6-9376-713c839b31f6)

**Comment by rkeithhill** [comment]
The Engine WG reviewed this issue and agrees with @Luiz-Monad. The message can easily be updated to display the int `ExitCode` in both decimal and hexadecimal.

**Comment by sba923** [comment]
@rkeithhill should I bite the bullet and create a PR for that?

**Comment by sba923** [comment]
This _should_ be a 3-line change 😜

![image](https://github.com/PowerShell/PowerShell/assets/12860484/2854b334-fc18-4b43-ab90-ab5030082414)

![image](https://github.com/PowerShell/PowerShell/assets/12860484/7f4a1d21-f058-471b-be5f-06608e47de1f)

BTW, where are the _localized_ strings for the error messages?

**Comment by rkeithhill** [comment]
LGTM!

**Comment by rkeithhill** [comment]
Can we get a quick review from someone on the PS team? I think this is a pretty straightforward, low-risk change.

**Comment by sba923** [comment]
How can progress be made on this? The PR is there and has been reviewed, right?

**Comment by sba923** [comment]
> How can progress be made on this? The PR is there and has been reviewed, right?

@rkeithhill how can we unblock this?

**Comment by rkeithhill** [comment]
@sba923 I'll try asking again. This is a pretty straightforward change.  Hopefully, we can get this in for 7.6.0.  🤞 @iSazonov and @daxian-dbw, any chance you could review the [associated PR](https://github.com/PowerShell/PowerShell/pull/21067)?  This a minimal change PR that I think would benefit PowerShell.

**Comment by iSazonov** [comment]
There were differences in formatting. My preference was and still is not to use leading zeros as indicated in [the documentation](https://learn.microsoft.com/en-us/windows/win32/debug/system-error-codes--0-499-). I have not seen any arguments as to why the WG wants them on Windows. Therefore, the process stopped.

**Comment by rkeithhill** [comment]
@iSazonov If that's the only thing holding up this PR, let's eliminate the leading zeros and get this PR merged.

**Comment by iSazonov** [comment]
I am participating in the discussion with the same rights as the others and I cannot overturn the decisions of the WG. (Even if I disagree, I merge such PRs.)
But Dongbo was assigned there. It would have been incorrect to replace it at that time. 
Now that so much time has passed without progress, I suggest closing this PR, rebasing it to get latest commits, and opening a new PR. If you agree to use the same format on all platforms, then specify this explicitly and fix the code. If not, then leave the code as it is. I will promote the new PR in any form.

**Comment by daxian-dbw** [comment]
Close via #21067
